### ISSUE TITLE
Make no-case-declarations rule fixable

### ISSUE THREAD

**Issue by medikoo** [eslint/eslint#8567]
When error is spotted, fix will be about wrapping `case` clauses with blocks. It'll be great to have it.

---

__What rule do you want to change?__

`no-case-declarations`

__Does this change cause the rule to produce more or fewer warnings?__

No

__How will the change be implemented? (New option, new default behavior, etc.)?__

_Fix_ functionality will be configured

__Please provide some example code that this change will affect:__

Code as:

```javascript
switch (foo) {
case "bar":
  let bar;
}
```

Will be fixed to:

```javascript
switch (foo) {
case "bar": {
  let bar;
}
}
```

__What does the rule currently do for this code?__

There's no _fix_ functionality configured

__What will the rule do after it's changed?__

_Fix_ functionality will be configured

**Comment by eslintbot** [comment]
Hi @medikoo, thanks for the issue. It looks like there's not enough information for us to know how to help you.

If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

Requesting a rule change? Please see [Proposing a Rule Change](http://eslint.org/docs/developer-guide/contributing/rule-changes) for instructions.

If it's something else, please just provide as much additional information as possible. Thanks!

**Comment by ljharb** [comment]
What about the more common indentation of:
```js
switch (foo) {
  case "bar": {
    let bar;
  }
}
```
(either the `case` line, or the let + `}` lines, were mis-indented in your fixed example, so i'm not sure what you were going for)

I would expect, ideally, the original indentation of the `case` to be maintained - such that:
```js
switch (x) {
case y:
  let z = x;
}

switch (a) {
  case b:
    let c = a;
}
```
would become:

```js
switch (x) {
case y: {
  let z = x;
}
}

switch (a) {
  case b: {
    let c = a;
  }
}
```

**Comment by medikoo** [comment]
@ljharb Sure indentation should be maintained, this fix should be purely about adding block wrap nothing more (I've updated example)

**Comment by not-an-aardvark** [comment]
Thanks for the proposal.

I don't think this should be fixable because the autofixer would end up changing the behavior of the code, which is generally something we try to avoid. For example, the following code would start throwing an error if a block was added:

```js
const x = 1;
switch (x) {
  case 1:
    const foo = 1;
  default:
    console.log(foo);
}
```

**Comment by medikoo** [comment]
@not-an-aardvark indeed we can't be sure whether that's intentional or not, therefore applying a fix may break working code. I didn't envision that when submitting this issue. I'm closing it then

**Comment by medikoo** [comment]
@not-an-aardvark giving this a second thought. Can you find a valid use case for writing code like in your example?
This code is not valid on it's own, as passing anything but `1` crashes the code, and writing `switch` purely for one possible case also doesn't make sense.

I take fixing with block wrap, won't be harmful for such cases, as they anyway could be only not intended.

**Comment by not-an-aardvark** [comment]
In general we try to avoid breaking working code even when the original code is poorly written. Also see: https://github.com/eslint/eslint/issues/7593

Another reason I'm concerned about autofixing this is that the `no-case-declarations` rule is highlighting a potential programmer error -- as you said, it's unlikely that the programmer intended trigger errors from this rule. However, I don't think we can be sure what the programmer intended to write. For example, perhaps they just misunderstand how block scoping works, and the variables are actually supposed to be declared outside of the `switch` statement entirely.

With an autofixer, the rule would be making a guess about what the programmer intended, and afterwards there would be no error reported, so the programmer might not even realize there was a problem. This could cause confusion if the rule makes an incorrect assumption about the programmer's intent. On the other hand, when there is no autofixer, the rule highlights the error and programmer can fix it appropriately. This is why most rules for highlighting potential errors don't have autofixes.

**Comment by medikoo** [comment]
@not-an-aardvark ok, I see, if the rule is simply "do not change the code behavior by any means" (which btw sounds very sane) then definitely fix should not be applied here (unconditionally).

Still, in this case we can detect whether variables are used outside the `case` clause and apply patch only if they doesn't? Would that be acceptable solution?

**Comment by platinumazure** [comment]
I'm neutral to allowing autofix only if a variable is used inside the `case` where it is declared. I don't think there would necessarily be a problem, but I also don't really see this as super valuable and worth the maintenance cost or risk. (Just my personal opinion.)

**Comment by not-an-aardvark** [comment]
Thanks for your interest in improving ESLint. Unfortunately, it looks like this issue didn't get enough support from the team and so I'm closing it. While we wish we'd be able to accommodate everyone's requests, we do need to prioritize. We've found that issues failing to reach consensus after a long time tend to never do it, and as such, we close those issues. This doesn't mean the idea isn't interesting, just that it's not something the team can commit to.
